pkg/controller/hosted/controller.go

```diff
@@ -221,44 +221,85 @@ func (r *ReconcileHosted) getHostedManifestWork(ctx context.Context, namespace,
 //   3. delete the klusterlet and managed kubeconfig manifest works
-func (r *ReconcileHosted) deleteManifestWorks(ctx context.Context, cluster *clusterv1.ManagedCluster, works, hostedWorks []workv1.ManifestWork) error {
+func (r *ReconcileHosted) deleteManifestWorks(
+	ctx context.Context,
+	cluster *clusterv1.ManagedCluster,
+	works, hostedWorks []workv1.ManifestWork) (
+	reconcile.Result, error) {
 	if (len(works) + len(hostedWorks)) == 0 {
-		return nil
+		return reconcile.Result{}, nil
 	}
 
 	if helpers.IsClusterUnavailable(cluster) {
 		// the managed cluster is offline, force delete all manifest works
-		return helpers.ForceDeleteAllManifestWorks(ctx, r.clientHolder.RuntimeClient, r.recorder, append(works, hostedWorks...))
+		return reconcile.Result{}, helpers.ForceDeleteAllManifestWorks(
+			ctx, r.clientHolder.RuntimeClient, r.recorder, append(works, hostedWorks...))
 	}
 
-	// delete works that do not include klusterlet addon works, the addon works will be removed by
-	// klusterlet-addon-controller, we need to wait the klusterlet-addon-controller delete them
+	// delete works that do not include klusterlet works and klusterlet addon works, the addon works were removed
+	// above, we need to wait them to be deleted.
 	ignoreAddons := func(clusterName string, manifestWork workv1.ManifestWork) bool {
-		return strings.HasPrefix(manifestWork.GetName(), fmt.Sprintf("%s-klusterlet-addon", manifestWork.GetNamespace()))
+		manifestWorkName := manifestWork.GetName()
+		switch {
+		case strings.HasPrefix(manifestWorkName, fmt.Sprintf("%s-klusterlet-addon", manifestWork.GetNamespace())):
+		case strings.HasPrefix(manifestWorkName, "addon-") && strings.HasSuffix(manifestWork.GetName(), "-deploy"):
+		case strings.HasPrefix(manifestWorkName, "addon-") && strings.HasSuffix(manifestWork.GetName(), "-pre-delete"):
+		default:
+			return false
+		}
+		return true
 	}
 	err := helpers.DeleteManifestWorkWithSelector(ctx, r.clientHolder.RuntimeClient, r.recorder, cluster, works, ignoreAddons)
 	if err != nil {
-		return err
+		return reconcile.Result{}, err
+	}
+
+	noAddons, err := helpers.NoManagedClusterAddons(ctx, r.clientHolder.RuntimeClient, cluster.GetName())
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if !noAddons {
+		// wait for addons deletion
+		return reconcile.Result{RequeueAfter: 10 * time.Second}, nil
 	}
 
 	ignoreNothing := func(_ string, _ workv1.ManifestWork) bool { return false }
 	noPending, err := helpers.NoPendingManifestWorks(ctx, r.clientHolder.RuntimeClient, log, cluster.GetName(), ignoreNothing)
 	if err != nil {
-		return err
+		return reconcile.Result{}, err
 	}
 	if !noPending {
 		// still have other works, do nothing
-		return nil
+		return reconcile.Result{}, nil
 	}
 
-	return r.deleteHostedManifestWorks(ctx, r.clientHolder.RuntimeClient, r.recorder, cluster)
+	return reconcile.Result{}, r.deleteHostedManifestWorks(ctx, r.clientHolder.RuntimeClient, r.recorder, cluster)
```

Comment:
should we also RequeueAfter here?

Reply:
I don't think it is necessary, if `deleteHostedManifestWorks` returns err, it will requeue. And the `deleteHostedManifestWorks` func only return get/delete manifestwork error which should be retried immediately

Reply:
what i mean is when you have pending works, should you return error which requeu immediately, or should it requeue after a certain time.

Reply:
if the work is pending, it will return nil. see [here](https://github.com/stolostron/managedcluster-import-controller/blob/cc6cab558ff07ea844bb0c106f53350f900b352e/pkg/helpers/manifestwork.go#L127)

Reply:
hrm, I think we should discuss about that, but it is not a blocker for this PR.